frontend/src/utils/hotkey_listener.ts

```diff
@@ -0,0 +1,220 @@
+    }
+  }
+
+  private bindKey(type: 'keyup' | 'keydown', keysOrOpts: KeysOrOpts, cb: KeyEventHandler) {
+    const { keys, scope, element } = this.forceOpts(keysOrOpts)
+    const scopedCallback = (event: KeyboardEvent, handler: HotkeysEvent) => {
+      if (this.selectedScopes.includes(scope)) {
+        return cb(event, handler)
+      }
+    }
+
+    if ( type === 'keyup' ) {
+      hotkeys(keys, { scope: 'all', element, keyup: true, keydown: false }, scopedCallback)
+    } else { // keydown
+      hotkeys(keys, { scope: 'all', element, keyup: false, keydown: true }, scopedCallback)
+    }
+
+
+    if (!this.scopeCallbackRegistry[scope]) {
+      this.scopeCallbackRegistry[scope] = []
+    }
+    this.scopeCallbackRegistry[scope].push({ keys, callback: scopedCallback })
+  }
+
+  onKeyup(keysOrOpts: KeysOrOpts, cb: KeyEventHandler) {
+    this.bindKey('keyup', keysOrOpts, cb)
+  }
+
+  onKeydown(keysOrOpts: KeysOrOpts, cb: KeyEventHandler) {
+    this.bindKey('keydown', keysOrOpts, cb)
+  }
+
+  addScope(scope: string) {
+    if (!this.selectedScopes.includes(scope)) {
+      this.selectedScopes.push(scope)
+    }
+  }
+
+  removeScope(scope: string) {
+    const index = this.selectedScopes.indexOf(scope)
+    if (index > -1) {
+      this.selectedScopes.splice(index, 1)
+    }
+  }
+
+  clear(scope?: string, keys?: string | string[]) {
+    // If neither scope nor keys are provided, clear all hotkeys
+    if (!scope && !keys) {
+      hotkeys.unbind()
+      this.scopeCallbackRegistry = {} // Reset the registry
+      return
+    }
+
+    // If only scope is provided, clear all hotkeys for that scope
+    if (scope && !keys) {
+      if (this.scopeCallbackRegistry[scope]) {
+        for (const callback of this.scopeCallbackRegistry[scope]) {
+          hotkeys.unbind(callback.keys, callback.callback)
+        }
+        delete this.scopeCallbackRegistry[scope]
+      }
+      return;
+    }
+
+    // If both scope and keys are provided, clear specific hotkeys within that scope
+    if (scope && keys) {
+      // Convert keys to array if it's a single string
+      if (typeof keys === 'string') {
+        keys = [keys]
+      }
+
+      // Filter callbacks for the specific scope and keys
+      const callbacksToRemove =
+        this.scopeCallbackRegistry[scope] && this.scopeCallbackRegistry[scope].filter(callback => keys.includes(callback.keys))
+
+      if (callbacksToRemove) {
+        callbacksToRemove.forEach(callback => {
+          hotkeys.unbind(callback.keys, callback.callback)
+        });
+      }
+    }
+  }
+
+  deleteScope(scope: string) {
+    this.clear(scope)
+    if (this.scopeCallbackRegistry[scope]) {
+      delete this.scopeCallbackRegistry[scope]
+    }
+
+    this.removeScope(scope)
+  }
+
+  /**
+   * Add a filter. The filter should return `true` if the event should be processed, and `false` otherwise.
+   * @param filter - The filter function to add.
+   */
+  addFilter(filter: (event: KeyboardEvent) => boolean) {
```

Comment:
Updated HotkeyListener to enable support for custom filters that prevent hotkey handlers (callbacks) from firing.